mri/reconstruct/fourier.py

```diff
@@ -208,3 +208,139 @@ def adj_op(self, x):
+        # This is done for easier stacking
+        self.sort_pos = np.lexsort(tuple(samples[:, i]
+                                         for i in np.arange(3)))
+        samples = samples[self.sort_pos]
+        plane_samples, self.z_samples = self.get_stacks(samples)
+        if fft_type == 'NUFFT':
+            self.FT = NUFFT(samples=plane_samples, shape=shape[0:2],
+                            platform=platform)
+        elif fft_type == 'NFFT':
+            self.FT = NFFT(samples=plane_samples, shape=shape[0:2])
+
+    def get_stacks(self, samples):
+        """ Function that converts an incoming 3D kspace samples
+            and converts to stacks of 2D. This function also checks for
+            any issues incoming k-space pattern and if the stack property
+            is not satisfied.
+            Stack Property:
+                The k-space locations originate from a stack of 2D samples
+        Parameters
+        ----------
+        samples: np.ndarray
+            the mask samples in the 3D Fourier domain.
+
+        Returns
+        ----------
+        plane_samples: np.ndarray
+            A 2D array of samples which when stacked gives the 3D samples
+        z_samples: np.ndarray
+            A 1D array of z-sample locations
+        """
+        self.first_stack_len = np.size(np.where(samples[:, 2]
+                                                == np.min(samples[:, 2])))
+        self.acq_num_slices = int(len(samples) / self.first_stack_len)
+        stacked = np.reshape(samples, (self.acq_num_slices,
+                                       self.first_stack_len, 3))
+        z_expected_stacked = np.reshape(np.repeat(stacked[:, 0, 2],
+                                                  self.first_stack_len),
+                                        (self.acq_num_slices,
+                                         self.first_stack_len))
+        if np.mod(len(samples), self.first_stack_len) \
+                or not np.all(stacked[:, :, 0:2] == stacked[0, :, 0:2]) \
+                or not np.all(stacked[:, :, 2] == z_expected_stacked):
+            raise ValueError('The input must be a stack of 2D k-Space data')
+        plane_samples = stacked[0, :, 0:2]
+        z_samples = stacked[:, 0, 2]
+        z_samples = z_samples[:, np.newaxis]
+        return plane_samples, z_samples
+
+    def op(self, data):
+        """ This method calculates Fourier transform.
+        Parameters
+        ----------
+        img: np.ndarray
+            input image as array.
+
+        Returns
+        -------
+        result: np.ndarray
+            Forward 3D Fourier transform of the image.
+        """
+        first_fft = np.fft.fftshift(np.fft.fft(np.fft.ifftshift(data, axes=2),
+                                               n=self.acq_num_slices,
+                                               norm="ortho"),
+                                    axes=2)
+        final = np.asarray([self.FT.op(first_fft[:, :, slice])
```

Comment:
`final` may not be the best variable name.

Reply:
Hmm, I was not sure what to give, Ill name is result? I am a bit clueless on this..

Reply:
well what is it?

Reply:
It is the **final** K-space data which is stacked. Hence the name **final** :P

Reply:
why not `stacked_kspace` then?